fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunnerInterface.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * FDBDatabaseRunnerInterface.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * <p>
+ * A context for running against an {@link FDBDatabase} with retrying of transient exceptions.
+ * </p>
+ *
+ * <p>
+ * Implements {@link #run} and {@link #runAsync} methods for executing functions in a new {@link FDBRecordContext} and returning their result.
+ * </p>
+ *
+ * <p>
+ * Implements {@link #close} in such a way that {@code runAsync} will not accidentally do more work afterwards. In particular, in the case
+ * of timeouts and transaction retries, it is otherwise possible for a whole new transaction to start at some time in the future and change
+ * the database even after the caller has given up, due to a timeout, for example.
+ * </p>
+ *
+ *
+ * <pre><code>
+ * try (FDBDatabaseRunner runner = fdb.newRunner()) {
```

Comment:
Hmm, it is a _little_ strange that this example doesn't actually refer to `FDBDatabaseRunnerInterface`.   But, I guess without breaking the existing interface this is what we need to do?